tsdb/db.go

```diff
@@ -856,6 +856,9 @@ func (db *DB) CompactHead(head *RangeHead) error {
 // compactHead compacts the given RangeHead.
 // The compaction mutex should be held before calling this method.
 func (db *DB) compactHead(head *RangeHead) error {
+	if db.opts.RetentionDuration > 0 && head.MaxTime()-head.BlockMaxTime() > db.opts.RetentionDuration {
```

Comment:
I would expect these checks to be inside `Compact()`, because we use this for forced compactions (in Cortex and I guess even Thanos?).

Additionally `head.MaxTime()-head.BlockMaxTime()` is almost always `-1` unless there is a sample incoming between those 2 calls, so this is mostly going to be `false` :). Which means, we need test for this.